technical-docs/2021-05-remote-restore-redesign.md

```diff
@@ -0,0 +1,110 @@
+
+
+## Restore Orchestration
+
+1. Get backup definition from storage backend
+![](img/K8ssandra-new-restore-design-1.jpg)  
+The restore operation will assume the CRDs have been synced with the remote storage backend. Backup topology (hostnames and tokens) will be retrieved from the new CRD format.
+
+
+2. Set the env variables on the restore containers
+![](img/K8ssandra-new-restore-design-2.jpg)  
+The backup definition will be used to map the backup hosts with the restore hosts and perform checks on whether or not the restore is possible:
+
+  - Same number of DCs
+  - Same number of nodes per DC
+  - Same rack distribution per DC
+  
+  Nodes will be mapped 1:1 between the source and destination topologies (sorted alphabetically). If the mapping fails to associate nodes 1:1, the restore operation must fail.
+
+  The init containers will get the following env variables set:
+
+  - Backup name
+  - Restore key
+  - Bucket prefix (new)
+  - In Place/Remote (new)
+
+  These values will then be used to invoke medusa.restore_node in the medusa-restore init container.
+The init-container needs to deal with mapping between source and dest topologies to self determine which backup it needs to restore. Pre-checks in the operator will guarantee that a 1:1 mapping is possible and topologies are compatible.
+
+3. Shutdown Cassandra to trigger the restore
+![](img/K8ssandra-new-restore-design-3.jpg)  
+
+  medusa-operator will patch the cassandradatacenter spec to provoke a shutdown of the Cassandra nodes through cass-operator. The pods will be restarted and the medusa-restore init container will trigger the restore.
+
+4. Restore the backups
+
+![](img/K8ssandra-new-restore-design-4.jpg)  
+
+  The medusa-restore container will perform a remote restore which involves the following:
+system.local and system.peers will not be restored
+auto_bootstrap will be set to false and seed nodes will start first
+tokens will be enforced by modifying cassandra.yaml
+
+  medusa-operator will monitor the restore by checking the cassandradatacenter object state.
+
+![](img/K8ssandra-new-restore-design-5.jpg)  
+
+  Once the cassandradatacenter is in Ready state, the restore will be marked as finished and the cluster will be fully restored and available.
+
+## Other required changes
+
+**The cassandradatacenter object will not be recreated during restore.**
+Currently the remote restore procedure assumes that the Cassandra cluster is not running and that the cassandradatacenter object hasn’t been created.
+This is a problem because creating the cassandradatacenter shouldn’t predate the helm install which includes other components (Reaper, monitoring, …) that will not be restored by Medusa.
+The new procedure will assume that a cassandradatacenter has been created already using the same number of nodes (token assignments can be different and will be enforced, datacenter name can differ). 
```

Comment:
If the new procedure assumes that the CassandraDatacenter already exists then would the restore procedure be the same or similar to what is done for in-place restores with respect to the restore controller?

Reply:
Not really, because remote restores do two things differently compared to in place restores:

- `system.local` and `system.peers` are not restored
- tokens are enforced and the nodes need to bootstrap